### ISSUE TITLE
Option#orNull can't be used from java

### ISSUE THREAD

**Issue by mosesn** [scala/bug#11625]
I raised this issue before https://github.com/scala/bug/issues/10006, but the issue wasn't well enough specified, so it was closed.  Let me try again.

Scala options are painful to use from java, and java users often want to simply use the actual value or a null, which is often idiomatic in java.  In particular, tools like findbugs can even make this kind of programming somewhat pleasant.  For java compatibility, it's very useful to be able to convert directly from an `Option` type to a "raw value or null" type.  Option#orNull seems like it should be able to do this, but it can't.  The bug here to me is that you can't unwrap your Option easily.

Here was the error message I got at the time:

```
error] class AllOf implements RoutePredicate {
                       [error] blahblaaablhablah.java:24: method orNull in class scala.Option<A> cannot be applied to given types;
                       [error]   required: scala.Predef.$less$colon$less<scala.runtime.Null$,A1>
                       [error]   found: no arguments
                       [error]   reason: cannot infer type-variable(s) A1
                       [error]     (actual and formal argument lists differ in length)
                       [error]     String sanitizedPath = opt.orNull();
                       [error] Compile failed at Oct 24, 2016 1:38:55 PM [0.221s]

               compile(blabhlahblahblahbl/predicates:predicates) failed: Zinc compile failed.
```

Someone suggested that users try `Option#getOrElse(null)` which will compile, but will result in a NPE if the `else` case is ever taken.  The correct way of implementing it would be `Option#getOrElse(() -> null)`, but this is easy to get wrong, allocates a function unnecessarily, and is surprising.

How about adding a method like `Option#unsafeOrNull` that doesn't ask for implicit evidence, and directly returns the contents of the Option in the Some case, and null in the None case?

**Comment by hrhino** [comment]
We could keep `orNull` as it is, but not use the evidence parameter, so
```scala
def orNull(implicit ev: Null<:<A): A =
  if (isDefined) get else null.asInstanceOf[A]
```
which would allow `opt.orNull(null)` I guess, if we're OK with providing a slightly crusty API to Java (as we already often do).

I don't like `unsafeOrNull`. Arguably the code above (without the evidence) could be `orZero`, because `Option.empty[Int].orZero` would be `0`, but they might not be obvious to most.

As a side note, `() -> null` should only allocate a single instance in Java, which will be cached at the callsite.

**Comment by Jasper-M** [comment]
It's not the most convenient thing if all you want is just the bare value or null, but there's https://www.scala-lang.org/api/2.13.0/scala/jdk/javaapi/OptionConverters$.html now.

**Comment by lrytz** [comment]
In larger mixed Scala-Java projects, there are maybe more issues of this kind, so I would expect there is some interop utility class. Is this a valid assumption? Writing a static `orNull` method in Java is quite simple.

I'm not saying we should ignore this issue, if there's a good way to fix it of course we have to do it. But the workaround doesn't seem too bad.

**Comment by ashawley** [comment]
> We could keep `orNull` as it is, but not use the evidence parameter

Wouldn't the implicit evidence parameter still cause it to not compile in Java?

> Arguably the code above (without the evidence) could be `orZero`, because `Option.empty[Int].orZero` would be `0`, but they might not be obvious to most.

This is because `Int` is a boxed primitive?  For Java reference types it would be `null`, e.g. `Option.empty[java.lang.Integer].orNull` returns null, today, in Scala.  Seems that would likely be the use case for Java client code.

**Comment by mosesn** [comment]
@lrytz yes, that's true, but it also increases the burden of getting ramped up in these codebases.  we want to expose APIs that are both scala-friendly and java-friendly, but when you encounter a scala Option and need to then go digging around for the appropriate helper to unwrap it for you, it's not very pleasant to use.

**Comment by som-snytt** [comment]
Converting to `Optional` requires no digging.

>  we want to expose APIs that are both scala-friendly and java-friendly

Previous opinion is that this is a pipe dream.

I wonder if documenting Java-(un)friendliness is a worthy goal.

**Comment by mosesn** [comment]
Thank you for looking at my ticket, but it was a bummer to return to this thread to see that the selling point of Scala libraries being interoperable from Java was described as a pipe dream.

**Comment by som-snytt** [comment]
I was disappointed to return to this ticket and find I don't remember my conclusions.

The conclusion was that `OptionConverters` works as shown in the PR test, but not that "most Java users are used to inconvenient APIs", apparently.

The trivial Scala shim is:
```
def orNull[A >: Null](x: Option[A]) = x.orNull
```
where the bound is invisible to Java.

The "previous opinion" was that if you want a stable, easy-to-use interface to use from Java, specify Java interfaces. (Akka.)

The doc example shows supplying null to Java from Scala, not conversely: 
`val textField = new JComponent(initialText.orNull,20)`

It's worth noting that there are several methods on `Option` requiring implicits.

I see that I tried out just supplying a value for the function, but found it challenging. FWIW, this works:
```
scala.Option<String> x;
class Z<T> extends scala.$less$colon$less<scala.runtime.Null$, T> {
        public <A> A substituteBoth(A a) { return a; }
}
public <A> Z<A> za() { return new Z<>(); }
String s = x.orNull(za());

public static final Z<String> z = new Z<>();
var y = x.<String>orNull(z); // I expected not to specify a type arg
```
